File: tests/source_samples/javascript/data_types/test_data_types.js
// JavaScript Data Types and Operations Test Suite

// Number and Math
let n1 = 10, n2 = 4;
n1 + n2;
n1 - n2;
n1 * n2;
n1 / n2;
n1 % n2;
n1 ** 2;
Math.sqrt(144);
Math.floor(3.7);
Math.ceil(3.2);
Math.round(3.5);
Math.abs(-5);

// String operations
let str1 = "hello";
let str2 = "world";
str1 + " " + str2;
str1.length;
str1[1]; // "e"
str1.substring(1, 4); // "ell"
str1.toUpperCase();
str2.toLowerCase();
"  spaced  ".trim();
"comma,separated".split(",");

// Array operations
let arr = [1, "two", 3.0];
arr.length;
arr.push(4);
arr.pop();
arr[0] = "one";
arr.map(x => typeof x);
arr.filter(x => typeof x === "number");
arr.reduce((acc, val) => acc + val, 0);

// Object operations
let obj = { name: "John", age: 30 };
obj.name;
obj['age'];
obj.city = "New York";
delete obj.age;
Object.keys(obj);
Object.values(obj);
Object.entries(obj);

// Type Coercion
1 == '1'; // true
1 === '1'; // false
'5' - 3; // 2 (number)
'5' + 3; // '53' (string)
Boolean(0); // false
Boolean(""); // false
Boolean("hello"); // true
Number("123"); // 123
String(123); // "123"

// Array methods
const numbers = [1, 2, 3, 4, 5];
numbers.forEach(n => n * 2);
numbers.map(n => n * 2);
numbers.filter(n => n > 2);
numbers.find(n => n > 3);
numbers.findIndex(n => n > 3);
numbers.some(n => n > 4);
numbers.every(n => n > 0);
numbers.includes(3);
numbers.indexOf(3);

// Object methods
const person = { name: "Alice", age: 25 };
const hasName = 'name' in person;
const hasOwn = person.hasOwnProperty('age');
const merged = Object.assign({}, person, { city: "NYC" });
const spread = { ...person, city: "NYC" };

console.log("Data types and operations test completed");

File: tests/source_samples/python/data_types/test_data_types.py
# Python Data Types and Operations Test Suite

# Numeric Types and Operations
a, b = 10, 3
a + b
a - b
a * b
a / b
a // b
a % b
a**b
c = -5
+c
-c
abs(c)
divmod(a, b)
pow(a, b)
round(3.14159, 2)

# String Type and Operations
s1 = "hello"
s2 = "world"
s1 + " " + s2
s1 * 3
"e" in s1
"x" not in s1
s1[1]
s1[-1]
s1[1:4]
len(s1)
"Hello".lower()
"world".upper()
"  spaced  ".strip()
"comma,separated".split(",")

# List Type and Operations
my_list = [1, "two", 3.0]
len(my_list)
my_list.append(4)
my_list.pop()
my_list[0] = "one"
nested_list = [1, [2, 3], 4]
nested_list[1][1]
list_comp = [x * x for x in range(5)]

# Tuple Type and Operations
my_tuple = (1, "two", 3.0)
len(my_tuple)
my_tuple[1]
try:
    my_tuple[1] = "new"
except TypeError:
    pass  # Expected - tuples are immutable
a, b, c = my_tuple

# Dictionary Type and Operations
my_dict = {"one": 1, "two": 2}
my_dict["one"]
my_dict["three"] = 3
"three" in my_dict
list(my_dict.keys())
list(my_dict.values())
del my_dict["two"]
"two" not in my_dict
dict_comp = {x: x * x for x in range(3)}

# Set Type and Operations
my_set = {1, 2, 2, 3}
my_set.add(4)
4 in my_set
my_set.remove(2)
2 not in my_set
set1 = {1, 2, 3}
set2 = {3, 4, 5}
set1 | set2
set1 & set2
set1 - set2
set1 ^ set2
set_comp = {x for x in "abracadabra" if x not in "abc"}

print("Data types and operations test completed")
